prompt2model/param_selector/search_with_optuna.py

```diff
@@ -0,0 +1,245 @@
+        best_model_path = "./best_model"
+        if not os.path.exists(best_model_path):
+            os.makedirs(best_model_path, exist_ok=True)
+
+        best_hyperparameters = self.optimize_hyperparameters(
+            training_datasets=training_datasets,
+            validation=validation,
+            hyperparameters=hyperparameters,
+        )
+        final_hyperparameters = {
+            "output_dir": "./best_model_checkpoint",
+            **best_hyperparameters,
+        }
+
+        model, tokenizer = self.generation_model_trainer.train_model(
+            hyperparameter_choices=final_hyperparameters,
+            training_datasets=training_datasets,
+        )
+
+        model.save_pretrained(best_model_path)
+        tokenizer.save_pretrained(best_model_path)
+        return model, tokenizer
+
+    def _build_hp_space(
+        self, hyperparameter_space: Optional[dict[str, Any]] = None
+    ) -> dict[str, Any]:
+        return (
+            {
+                "min_weight_decay": hyperparameter_space.get(
+                    "min_weight_decay",
+                    DEFAULT_HYPERPARAMETERS_SPACE.get("min_weight_decay"),
+                ),
+                "max_weight_decay": hyperparameter_space.get(
+                    "max_weight_decay",
+                    DEFAULT_HYPERPARAMETERS_SPACE.get("max_weight_decay"),
+                ),
+                "min_learning_rate": hyperparameter_space.get(
+                    "min_learning_rate",
+                    DEFAULT_HYPERPARAMETERS_SPACE.get("min_learning_rate"),
+                ),
+                "max_learning_rate": hyperparameter_space.get(
+                    "max_learning_rate",
+                    DEFAULT_HYPERPARAMETERS_SPACE.get("max_learning_rate"),
+                ),
+                "min_num_train_epochs": hyperparameter_space.get(
+                    "min_num_train_epochs",
+                    DEFAULT_HYPERPARAMETERS_SPACE.get("min_num_train_epochs"),
+                ),
+                "max_num_train_epochs": hyperparameter_space.get(
+                    "max_num_train_epochs",
+                    DEFAULT_HYPERPARAMETERS_SPACE.get("max_num_train_epochs"),
+                ),
+                "save_strategy": hyperparameter_space.get(
+                    "save_strategy",
+                    DEFAULT_HYPERPARAMETERS_SPACE.get("save_strategy"),
+                ),
+                "evaluation_strategy": hyperparameter_space.get(
+                    "evaluation_strategy",
+                    DEFAULT_HYPERPARAMETERS_SPACE.get("evaluation_strategy"),
+                ),
+                "per_device_train_batch_size": hyperparameter_space.get(
+                    "per_device_train_batch_size",
+                    DEFAULT_HYPERPARAMETERS_SPACE.get("per_device_train_batch_size"),
+                ),
+            }
+            if hyperparameter_space is not None
+            else DEFAULT_HYPERPARAMETERS_SPACE
+        )
```

Comment:
I wonder if this can't be condensed programatically, since each of these 8 blocks does exactly the same thing. You would just have a constant that specifies the fields to populate.

Reply:
We can condense this with a for loop I guess?

Reply:
Going for that.